types/decimal.go

```diff
@@ -0,0 +1,324 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const Precision = 10
+
+func precisionInt() *big.Int {
+	return new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// get the precision multiplier
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
+	}
+	zerosToAdd := Precision - prec
+	multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	return multiplier
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec !> Precision
```

Comment:
should the contract be: `prec <= Precision` or `Precision > prec`, the `!>` looks odd to me?

Reply:
sure, updated to `prec <= Precision`